### PR TITLE
Support playing videos

### PR DESCRIPTION
**Pull request by marco-c**

### CODE REVIEW

**Comment by marco-c** [comment]
Probably Travis doesn't have the necessary codecs to decode the h264 video. I think we'd need sudo to install GStreamer via a package manager and sudo:true wouldn't allow us to use Travis caching.

**Comment by marco-c** [comment]
Yes, installing GStreamer fixes the problem. So, I think our only options are enabling the tests and disabling caching or disabling the tests and keep caching.

**Comment by marco-c** [comment]
Looks like Travis is eventually going to switch to the contained-based infrastructure (where caching is enabled), even when sudo:true.
This means we're going to lose caching only until Travis fixes their problems with sudo:true.

File: midp/media.js [hunk 1]
```diff
@@ -385,6 +388,121 @@ ImagePlayer.prototype.setVisible = function(visible) {
...
+                fs.close(fd);
+                this.video.src = URL.createObjectURL(new Blob([ videoData ]),
+                                                     { type: this.playerContainer.contentType });
+            }).bind(this));
+        } else {
+            this.video.src = this.playerContainer.url;
+        }
+    }).bind(this));
+
+    p.done(function() {
```
Comment (mykmelez) [inline midp/media.js]: Erm, does Gecko's Promise implementation support *done*? The polyfill may do so, but I don't think the Gecko one does.

File: midp/media.js [hunk 2]
```diff
@@ -385,6 +388,121 @@ ImagePlayer.prototype.setVisible = function(visible) {
...
+    this.video.style.left = x + "px";
+    this.video.style.top = y + "px";
+    this.video.style.width = w + "px";
+    this.video.style.height = h + "px";
+    document.getElementById("main").appendChild(this.video);
+}
+
+VideoPlayer.prototype.setVisible = function(visible) {
+    this.video.style.visibility = visible ? "visible" : "hidden";
+    if (this.isPlaying) {
```
Comment (mykmelez) [inline midp/media.js]: Shouldn't this also check that *visible* is `true`? Seems like we shouldn't start playing the video if it's `false`, even if *this.isPlaying*.

Comment (marco-c) [inline midp/media.js]: Done in 3bbdd3b (play is ignored if the element is hidden, so even without this change we wouldn't have started the video, but it's clearer this way).
There might be other things that we need to do when a MIDlet modifies the player's visibility several times, but it's hard to tell until we find a MIDlet that uses the video player in a more complicated way.

**Comment by mykmelez** [comment]
How do we account for developers who don't have the test dependencies on their system? I'm not sure I want to require them to install gstreamer plugins and ffmpeg just to run tests. Can we disable the test if the dependencies aren't available?

**Comment by mykmelez** [comment]
FWIW, I can `brew install gstreamer ffmpeg` on Mac, but I still get a test failure on this branch:

> 1:1  | javax.microedition.media.TestVideoPlayer-0: fail [Unexpected exception: javax.microedition.media.MediaException: Failed to load video] 
> 1:1  | javax.microedition.media.MediaException: Failed to load video 
> 1:1  |  - javax/microedition/media/MediaException.<init>(), bci=0 
> 1:1  |  - com/sun/mmedia/PlayerImpl.realize(), bci=0 
> 1:1  |  - javax/microedition/media/TestVideoPlayer.test(), bci=0 
> 1:1  |  - RunTests.runTest(), bci=0 
> 1:1  |  - RunTests.startApp(), bci=0 
> 1:1  |  - javax/microedition/midlet/MIDletTunnelImpl.callStartApp(), bci=0 
> 1:1  |  - com/sun/midp/midlet/MIDletPeer.startApp(), bci=0 
> 1:1  |  - com/sun/midp/midlet/MIDletStateHandler.startSuite(), bci=0 
> 1:1  |  - com/sun/midp/main/AbstractMIDletSuiteLoader.startSuite(), bci=0 
> 1:1  |  - com/sun/midp/main/CldcMIDletSuiteLoader.startSuite(), bci=0 
> 1:1  |  - com/sun/midp/main/AbstractMIDletSuiteLoader.runMIDletSuite(), bci=0 
> 1:1  |  - com/sun/midp/main/MIDletSuiteLoader.main(), bci=0 
> 1:1  |  - java/lang/Object.RaiseExceptionSynthetic(), bci=2 
> 1:1  |  - com/sun/mmedia/PlayerImpl.realize(), bci=76 
> 1:1  |  - javax/microedition/media/TestVideoPlayer.test(), bci=135 
> 1:1  | javax.microedition.media.TestVideoPlayer: 0 pass, 1 fail, 0 known fail, 0 unknown pass 
> 1:1  | javax.microedition.media.TestVideoPlayer: test expected 3 passes, got 0 
> 1:1  | javax.microedition.media.TestVideoPlayer: test expected 0 failures, got 1

**Comment by marco-c** [comment]
Mmmh, I thought on Mac Firefox was using the OS decoder. GStreamer should be used only on Linux.

**Comment by marco-c** [comment]
What do you see if you open index.html?main=com/sun/midp/main/MIDletSuiteLoader&midletClassName=gfx/VideoPlayerTest&jars=tests/tests.jar in Nightly?

**Comment by mykmelez** [comment]
> What do you see if you open index.html?main=com/sun/midp/main/MIDletSuiteLoader&midletClassName=gfx/VideoPlayerTest&jars=tests/tests.jar in Nightly?

I don't see an error, but I also don't see a video, nor does "PAINTED" get logged to the console.

**Comment by marco-c** [comment]
Maybe the video I've generated is somehow corrupted (I can't open it on Android), I'll test myself tomorrow (with another video as well).

**Comment by marco-c** [comment]
I'm using a WebM video for the test. Even if WebM videos are not actually used in J2ME MIDlets, the test is still complete (since our VideoPlayer implementation doesn't depend on the format of the video).

**Comment by marco-c** [comment]
Tests are passing both on Linux and Mac.

**Comment by mykmelez** [comment]
For posterity, to `make -C tests javax/microedition/media/test.webm` on Mac you need to first install ffmpeg with libvpx via something like `brew install ffmpeg --with-libvpx`.

**Pull request merged**